### PR TITLE
Move Encoding to be composed by gateway

### PR DESCRIPTION
**Pull request by neonphog**
## Review checklist 
- [ ] The story has unit or integration tests
- [ ] No new bugs, and any tech-debt is identified and justified
- [ ] There is enough API documentation (how to use)
- [ ] There is enough code documentation (how the code works)

### CODE REVIEW

File: crates/lib3h/src/gateway/gateway_dht.rs [hunk 1]
```diff
@@ -73,13 +72,21 @@ impl P2pGateway {
...
                 // Send phony SendMessage request so we connect to it
+                /*
                 self.inner_transport.publish(
                     span.follower("DhtRequestToParent::HoldPeerRequested"),
                     transport::protocol::RequestToChild::SendMessage {
                         uri: peer_data.peer_uri,
                         payload: Opaque::new(),
                     },
                 )?;
+                */
```
Comment (struktured) [inline crates/lib3h/src/gateway/gateway_dht.rs]: Can we delete this comment block?

Comment (neonphog) [reply crates/lib3h/src/gateway/gateway_dht.rs]: woops! meant to delete that

File: crates/lib3h/src/gateway/mod.rs [hunk 2]
```diff
@@ -30,6 +31,10 @@ pub struct P2pGateway {
     /// DHT
     inner_dht: Detach<ChildDhtWrapperDyn<P2pGateway>>,
 
+    /// message encoding actor
+    #[allow(dead_code)]
```
Comment (neonphog) [inline crates/lib3h/src/gateway/mod.rs]:
```suggestion
```

File: crates/lib3h/src/gateway/gateway_transport.rs [hunk 3]
```diff
@@ -77,22 +77,113 @@ impl P2pGateway {
...
+        let e_span = span.child("encode_payload");
+        self.message_encoding.request(
+            span,
+            encoding_protocol::RequestToChild::EncodePayload { payload },
+            Box::new(move |me, resp| {
+                match resp {
+                    GhostCallbackData::Response(Ok(
+                        encoding_protocol::RequestToChildResponse::EncodePayloadResult { payload },
+                    )) => {
+                        //error!("sending: {:?}", payload);
```
Comment (neonphog) [inline crates/lib3h/src/gateway/gateway_transport.rs]:
```suggestion
                        trace!("sending: {:?}", payload);
```

File: crates/lib3h/src/message_encoding.rs [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+            endpoint_parent,
+            endpoint_self,
+        }
+    }
+
+    fn handle_msg_from_parent(
+        &mut self,
+        mut msg: MessageEncodingMessageFromParent,
+    ) -> Lib3hResult<()> {
+        //error!("{:?}", msg.backtrace());
```
Comment (struktured) [inline crates/lib3h/src/message_encoding.rs]: Uncomment out or delete.

File: crates/lib3h/src/message_encoding.rs [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+        msg: MessageEncodingMessageFromParent,
+        payload: Opaque,
+    ) -> Lib3hResult<()> {
+        let payload = match InterimEncodingProtocol::from_slice(&payload) {
+            InterimEncodingProtocol::Handshake {
+                magic,
+                network_id,
+                id,
+            } => {
+                if magic != 0x1f6c {
```
Comment (struktured) [inline crates/lib3h/src/message_encoding.rs]: This is way too magical of a number. Can we give this meaningful constant name?

File: crates/lib3h/src/message_encoding.rs [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+                    msg.respond(Err("bad magic".into()))?;
```
Comment (struktured) [inline crates/lib3h/src/message_encoding.rs]: Any other useful info to push into the error message here?

File: crates/lib3h/src/message_encoding.rs [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+    }
+
+    fn handle_encode_handshake(
+        &mut self,
+        msg: MessageEncodingMessageFromParent,
+        space_address: String,
+        id: String,
+    ) -> Lib3hResult<()> {
+        let payload = InterimEncodingProtocol::Handshake {
+            magic: 0x1f6c,
```
Comment (struktured) [inline crates/lib3h/src/message_encoding.rs]: See comment above

File: crates/lib3h/src/message_encoding.rs [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+    fn take_parent_endpoint(&mut self) -> Option<MessageEncodingParentEndpoint> {
+        std::mem::replace(&mut self.endpoint_parent, None)
+    }
+
+    fn process_concrete(&mut self) -> GhostResult<WorkWasDone> {
+        detach_run!(&mut self.endpoint_self, |es| es.process(self))?;
+        for msg in self.endpoint_self.as_mut().drain_messages() {
+            self.handle_msg_from_parent(msg).expect("no ghost errors");
+        }
+        Ok(false.into())
```
Comment (struktured) [inline crates/lib3h/src/message_encoding.rs]: Is it worth honoring the `DidWork` contract here? Should be easily implementable.

File: crates/lib3h/src/message_encoding.rs (hunk 4 above)
Comment (neonphog) [inline crates/lib3h/src/message_encoding.rs]:
```suggestion
```

File: crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs [hunk 9]
```diff
@@ -272,6 +272,10 @@ impl
                     })))?;
                 }
                 RequestToChild::SendMessage { uri, payload } => {
+                    //error!("mem send: {:?}", payload);
+                    if payload[0] != b'{' {
```
Comment (struktured) [inline crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs]: Delete this?

**Pull request merged**